### ISSUE TITLE
option to disable PATH injection

### ISSUE THREAD

**Issue by glyph** [d12frosted/homebrew-emacs-plus#578]
My PATH changes from time to time depending on versions of things in my development environment, certain test features, etc.

I was *extremely* surprised to find the PATH injection in emacs-plus after spending like an hour trying to debug why my existing automation for synchronizing my environment and selecting e.g. Python versions was all broken.  Emacs was defaulting to using a non-framework build of Python 3.7 because that happened to be in my shell at the time that I did a `brew update`, even though I'd explicitly excluded that path.

This is just extremely counter to the way that apps generally work; I know that Homebrew isn't really for building redistributable bundles, but sticking paths from one specific machine into Info.plist is not how it's supposed to work.  I love a bunch of the features emacs-plus provides but I *super* do not want this.  I've deleted it out of my own Info.plist for now but I know it's going to come back in a few months.

Emacs *not* having your PATH on it is actually a tool I use to teach people about the nature of LaunchServices and the subtle ways your environment can vary.  Fixing this in this one build of emacs sets people up for weird expectations of other software if they switch to other editors or even other builds of Emacs.

There's also the issue of other users on the same machine getting garbled paths that reference different users' development environments.

If you really want a feature like this that is part of the homebrew build, could I suggest a patch that reads a `~/Library/Application Support/io.github.d12frosted.Emacs.environment.plist` or something, and dump that out during creation?  The idea of a mutable Info.plist that causes this kind of action-at-a-distance behavior makes my brain itch :).

**Comment by glyph** [comment]
> Hey @glyph
> 
> Sorry to hear that you had to spent a lot of time trying to figure out what's going on with your PATH. I totally understand that it's easy to miss the message about PATH injection in the build log.

> And thank you for sharing your thoughts, problems and proposing a workaround/alternative ❤️ I appreciate the way you approach the issue 💯

Hi @d12frosted. Thanks for understanding!

> TL;DR I would love to hear more about your workflow. But if needed, we can provide an option to opt out from PATH injection.
> 
> That being said, I want to better understand your workflow and why it's broken. So let me ask some questions.
> 
> 1. Can you please explain, what exactly you mean by 'PATH changes from time to time'? How exactly does it change (or rather where)?

As I move from project to project or my most common set of dependencies changes, I typically add and remove entries to a custom section in a file sourced by my `~/.zshrc`. Most commonly what I'm doing is adding/removing or asdf  entries based on the typical compatibility profile of most of my current projects.  I have somewhere between 20-30 entries on `$PATH` in a shell at any given time.

> 2. How do you start Emacs?

I typically launch it from the GUI. (Specifically, usually from Alfred, but not exclusively.)  When I want to inherit a *specific* shell environment globally because I'm doing something unusual, I launch it from the command line.

> > that happened to be in my shell at the time that I did a brew update, even though I'd explicitly excluded that path.
> 
> What do you mean by 'excluded that path'?

I edited the file I described above and deleted the entry I no longer wanted.

> > This is just extremely counter to the way that apps generally work; I know that Homebrew isn't really for building redistributable bundles, but sticking paths from one specific machine into Info.plist is not how it's supposed to work.
> 
> One the one hand, I understand what you mean. On the other hand, the way LaunchServices starts applications (e.g. no way to start from the interactive/custom environment - at least to my knowledge)

If you want Emacs to inherit a path from your shell, all you have to do is launch it from your shell; i.e., if you do `FOO=test-me open -a Emacs` at a shell when Emacs is not running, then `(getenv "FOO")` / <kbd>C-j</kbd> in your scratch buffer, it evaluates to `"test-me"`.

> is problematic for Emacs users (see [this blog post](https://d12frosted.io/posts/2022-05-12-emacs-plus-path-injection.html) for some reasoning). 

Yes, I am … far more familiar with this problem than I want to be, really :).  I developed my own alternative to `exec-path-from-shell` that, among other things, supports Windows, in order to accommodate my own esoteric workflow.  On Linux I have shell startup stuff that detects GUI sessions and specifically avoids setting env vars so that I do not get unintentional bleed-over from my shell to that environment.  The ability to quickly test in a clean, "known-good" environment without the hassle of a whole separate VM is important to me.

> And actually, not only for Emacs, but for other interactive environments with unmanaged external runtime dependencies (mostly editors).

To be clear here, what you mean by "unmanaged" is "managed by a shell".

> > Emacs not having your PATH on it is actually a tool I use to teach people about the nature of LaunchServices and the subtle ways your environment can vary.
> 
> IMO teaching != real world usage. The fact that you have to explain that topic to properly use various tools inside Emacs is just a proof (at least to me) that PATH injection is a good default.

I can appreciate that "teaching" is a somewhat esoteric case, but let me give you an example of a relevant situation where precise understanding is really useful.

A couple of users, A and B, with Emacs experience from introductory CS courses at their university start working together on an open source project with a custom toolchain.  A installs emacs-plus, then their dev tools. B installs the same things in the reverse order.

Because the dev tools edit `~/.zshrc` to make adding to `$PATH` easy, it works half the time here.

You've now created a situation where B is haivng a fine time, emacs-plus "just worked" for them by doing this incredibly obscure `$PATH` hack that just *happened* to have the correct `$PATH` value at the time they did `brew install` or `brew update`.  If they are trying to help A debug their environment and get set up properly, they might uninstall and then reinstall to get to a clean state, and then notice that their tools all broke, but they won't immediately know why unless they are vigilant for the post-install message (which, if it happens to be on a long command line like `brew install emacs-plus && brew install lots-of-tools && curl https://example.github.io/my-project/get-set-up.sh | sh`, they'll never notice).

Even more fun, put these users into different VNC sessions on the same CI server, and watch them absolutely howl in confusion as the error is not "command not found" but `/Users/a/dev/bin/custom-tool: permission denied` when B tries to run something because A was the one who `brew install`ed emacs.

> Though probably that calls for an option to opt out. It just that up to this day I found no real world use case where PATH injection is problematic. Teaching macOS launch system is something brought up for the first time 😄

I guess you work on a relatively small, fixed number of languages and tools? :) Installing new versions of Python, Rust, and Node all the time, I do not expect PATH to be the same from day to day, let alone the same from one major Emacs release to the next :).

Another thing to note is that PATH does not live in isolation. Some environments require a paired `DYLD_LIBRARY_PATH`, or perhaps options like `LDFLAGS`/`CPPFLAGS`/`CFLAGS`, to function properly.  If you set some but not the others, or set them inconsistently because the different values are coming from different versions (perhaps you're migrating a project between SDL1 and SDL2, and it's important to keep SDL1 off the path while you're testing your SDL2 build and vice versa?) then tools and builds break in truly odd ways.  Better to let users debug a simple error of the value not being set and needing to write a little elisp than to set *half* the values that they need, mysteriously forgetting the other half.

> > Fixing this in this one build of emacs sets people up for weird expectations of other software if they switch to other editors or even other builds of Emacs.
> 
> In general, I agree with you. At the same time, I think it should not stop me from providing a better UX for Emacs+ users.

All of the stuff I described above is pretty bad UX :).

> > could I suggest a patch that
> 
> Can you please elaborate more on that idea? I don't fully understand it.

1. Dump the environment variables for the user somewhere in their *home directory* rather than into the app bundle. i.e., somewhere typically considered "writable" by non-admin users.  The app bundle should be the same for everybody.  You can just stick this in `~/.emacs-plus.d/path-var.json` if you like.
2. Add a script that will do this so it can be updated post-facto without reinstalling and rebuilding emacs-plus.
3. Very early in startup, add a little C code to read that file and call `setenv` a few times before running any other bits of emacs.

**Comment by glyph** [comment]
> * Dump the environment variables for the user somewhere in their _home directory_ rather than into the app bundle. i.e., somewhere typically considered "writable" by non-admin users.  The app bundle should be the same for everybody.  You can just stick this in `~/.emacs-plus.d/path-var.json` if you like.

Oh another thing that I prefer about this approach: in addition to being more in line with the way things usually work, it's more discoverable (users can see the file) and easier to opt out of (a simple `rm` does it, rather than needing to know how to edit `Info.plist`)

**Comment by nidan841g** [comment]
> This is just extremely counter to the way that apps generally work; …

Automator is a MacOS app that can run a command in a shell. I can't think of another at the moment. I just ran `env` in Automator's shell script action. It had this `PATH` baked in:

    PATH=/usr/bin:/bin:/usr/sbin:/sbin

Is that what you want emacs-plus to do? That is definitely not the PATH that I have in my shells on my mac.

This is an old issue on Macs. [stackoverflow post from 11 yrs ago](https://stackoverflow.com/questions/8606954/path-and-exec-path-set-but-emacs-does-not-find-executable). There's a [page on the Emacs Wiki](https://www.emacswiki.org/emacs/ExecPath) that discusses it, and shows some elisp to set `exec-path` from a shell of your choosing.

**Comment by glyph** [comment]
> Automator is a MacOS app that can run a command in a shell. I can't think of another at the moment. I just ran `env` in Automator's shell script action. It had this `PATH` baked in:
> 
> ```
> PATH=/usr/bin:/bin:/usr/sbin:/sbin
> ```
> 
> Is that what you want emacs-plus to do?

I think two things that might make sense just to keep Emacs itself functional would be to include its own homebrew `bin` directory on `$PATH`, or perhaps even all of Homebrew, if it depends on binaries in that installation *itself* to function, such as `gnutls`, `pkg-config`, `etags` etc. But that doesn't come from the shell's `$PATH`, it's an integration built into Emacs itself for its own purposes.

> That is definitely not the PATH that I have in my shells on my mac.

That's only because specific shells have specific startup code that they run to set up your `$PATH`.  zsh, bash, and tcsh will have different paths if you run them separately, as will `fish` if you have it installed.

Go into Terminal, go to "New Command", and run "`sh --norc --noprofile`", you'll see a shell with only `/usr/bin:/bin` on `$PATH`.

Here's another thing that can easily happen: I use `virtualenvwrapper` which means that sometimes I do something like run the shell function `mktmpenv` and have an addition to `$PATH` in the current shell that's totally random and temporary, like `/Users/glyph/.virtualenvs/tmp-81e5831a880f224/bin`.  If I were to `brew install` a tool like that, it'll put some garbage on my path which will at best be ignored, and at worst will have Emacs hard-code a very confusing temporary tool environment with the wrong version of Python to the front of my path.

> This is an old issue on Macs. [stackoverflow post from 11 yrs ago](https://stackoverflow.com/questions/8606954/path-and-exec-path-set-but-emacs-does-not-find-executable). There's a [page on the Emacs Wiki](https://www.emacswiki.org/emacs/ExecPath) that discusses it, and shows some elisp to set `exec-path` from a shell of your choosing.

It's not exclusive to macs, though; consider [this question from 6 years ago](https://superuser.com/questions/1149622/gnome-3-22-how-to-change-the-path-for-the-shell). GUI environments do not necessarily run a shell's initialization code, and even if they do, they don't always run *your* shell's initialization code.

**Comment by d12frosted** [comment]
I am still not convinced that disabling path injection is a good idea. Users that don't manipulate their environment on per-directory basis benefit from path injection and users with complex use-cases still have to manually manipulate PATH inside Emacs or start Emacs in a way that environment is inherited.

**Comment by glyph** [comment]
> I am still not convinced that disabling path injection is a good idea. Users that don't manipulate their environment on per-directory basis benefit from path injection and users with complex use-cases still have to manually manipulate PATH inside Emacs or start Emacs in a way that environment is inherited.

Could I at least convince you to put this path injection into elisp somewhere, rather than the binary, so it can be more easily detected and adjusted by lisp code?

For the time being I guess I'm just going to work on a personal workflow where my `brew upgrade` environment is more aggressively protected from my personal / temporary path customizations.

**Comment by d12frosted** [comment]
@glyph I understand your concerns, but I won't be modifying Emacs sources.

While I see how the current path injection approach might be frustrating, the primary issue I acknowledge is the non-deterministic and non-reproducible nature of the hack. Despite this, it's still the best solution available.

When it comes to reproducibility, it's important to recognise that this is an inherent challenge with Homebrew. There are other package managers that place a greater emphasis on reproducibility.

So, what does this hack actually do? It allows `Emacs.app` to be run via a launch daemon on macOS, achieved through the modification of `Info.plist` file - not by injecting anything into the binary. This is the least intrusive method I could find, and I've tested it to ensure that the `PATH` value injected by Emacs+ can be easily replaced. I've documented some of these tests/my thoughts in #534. Additionally, this hack does mimic the behaviour seen in Linux distributions, at least based on my testing.

For other cases, I believe they must be handled using the appropriate tools, regardless of Emacs+ path injection. For example, you might use `exec-path-from-shell` or run Emacs from a specific environment, either via `open -na Emacs.app` (using the `-n` option to spawn a new instance) or by running the `emacs` binary directly. But you know that already. Some of the issues raised seem to stem more from Homebrew's limitations as a multi-user package manager, where Emacs+ is not the primary concern IMO.

To summarise, I believe the `Info.plist` path injection is a sufficient solution, addressing the majority of problems. Esoteric use cases requiring a dynamic environment should be addressed through other means, and path injection has no impact those scenarios. Even more specialised use cases are not something I'm willing to accommodate at the expense of most users. Advanced users should have no trouble figuring out these situations. For visibility, the Emacs+ formula includes additional logging to warn users about the injected `PATH` - though I acknowledge it might be easily overlooked.

Lastly, I don't plan to introduce a `--without-path-injection` option because I don't want to regress to the issues we faced before path injection was implemented. Too often, users follow installation instructions without fully understanding them, and if someone recommends `--without-path-injection`, we'd end up back where we started.

I hope this explanation makes sense. To avoid being overly rigid, here's an alternative solution: If you have installation scripts, you can simply add the following script to run after the Emacs+ installation.

```bash
EMACS_PLUS_PLIST="$(brew --cellar emacs-plus@30)/30.0.50/Emacs.app/Contents/Info.plist"

# delete PATH entry
/usr/libexec/PlistBuddy -c 'Remove :LSEnvironment:PATH' "$EMACS_PLUST_PLIST"

# optionally add some entries
/usr/libexec/PlistBuddy -c 'Add :LSEnvironment:PATH string' "$EMACS_PLUST_PLIST"
/usr/libexec/PlistBuddy -c "Set :LSEnvironment:PATH $PATH_VALUE"
```

Hope that helps.
```

**Comment by d12frosted** [comment]
> You might want to replace
Indeed, a typo 😅 Thanks for noticing.

> would you be amenable a PR which migrates to the alternate solution I proposed which would write a file into the user's home directory?

I can't say I like the proposed solution. It sounds good on paper, but it has several drawbacks. Keep in mind that I want this behaviour by default, which means some questions are tricky to answer. When to run the code that reads and sets PATH variables? What about startup performance penalty (even tiny)? How to make it non-obtrusive? How and when to update json? How to teach? And another point - I don't want Emacs+ to modify GNU Emacs (despite the fact that we have various patches - some of them are mistake).

IMO this kind of stuff (what you describe, `exec-path-from-shell` or `doom env`) should be invoked from configurations. And looking at `doom env` it seems that it's not that easy to implement without issues. All things considered I think PATH injection provided by Emacs+ is a decent solution - it's not ideal, maybe it's not even hygienic - but it works with very little drawbacks and almost zero maintenance.

> The way that Linux distributions are getting this sort of behavior is not to do something to the executable

I know. What I meant - it kind of 'emulates' result. Of course there are corner cases where the result with Emacs+ is plain wrong, but again - complex envs need to rely on a proper solution in macos anyways, so I don't care.

**Comment by glyph** [comment]
Ah, I think I was missing a key detail here: `LSEnvironment` is processed by macOS as a part of the app-launch process, rather than having *any* supporting code in Emacs+ itself.  I'd been assuming there was a stub somewhere I hadn't seen to process this during startup.  I actually found this out by looking around to discover this: https://attack.mitre.org/techniques/T1647/ which I am surprised is still listed as a malware technique, but I guess the fact that it has been listed as such for several years and *hasn't* even been deprecated yet (c.f. https://developer.apple.com/documentation/bundleresources/information_property_list/lsenvironment ), I suppose it's likely to be robust for longer than i would have guessed :).

I still don't love it but it makes more sense why you're treating `Info.plist` as a configuration file, since it's really the OS treating it that way not you. I still suspect that this will break at some point, but maybe that'll be in another 10 years, and we will have more pressing concerns by then.

**Comment by d12frosted** [comment]
@glyph yes, it's not unique to Emacs+, that's why I kept focusing on launch daemon and not Emacs+. Also this is something I explain in [the blog post](https://d12frosted.io/posts/2022-05-12-emacs-plus-path-injection.html) I've linked several times 😸 I also explain some other solutions for the problem in macos. 

> treating Info.plist as a configuration file, since it's really the OS treating it that way not you
Exactly.

> I suppose it's likely to be robust for longer than i would have guessed
> [...]
> I still suspect that this will break at some point, but maybe that'll be in another 10 years, and we will have more pressing concerns by then.

I doubt it will be changed in foreseeable future. But even if it does, we will adapt 😄 

I still do understand drawbacks/shortcomings of the taken approach, but IMO it's good enough all things considered.
